src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,678 @@
+        ret.p_max.y = OMT_SIZE + NAV_MAP_PADDING;
+    } else {
+        ret.p_min.y = OMT_SIZE + NAV_MAP_PADDING;
+        ret.p_max.y = OMT_SIZE + 2 * NAV_MAP_PADDING;
+    }
+    return ret;
+}
+
+static half_open_rectangle<point> compute_nav_zone( const half_open_rectangle<point> &goal_zone )
+{
+    half_open_rectangle<point> ret;
+    ret.p_min.x = std::min( goal_zone.p_min.x, NAV_MAP_PADDING );
+    ret.p_min.y = std::min( goal_zone.p_min.y, NAV_MAP_PADDING );
+    ret.p_max.x = std::max( goal_zone.p_max.x, OMT_SIZE + NAV_MAP_PADDING );
+    ret.p_max.y = std::max( goal_zone.p_max.y, OMT_SIZE + NAV_MAP_PADDING );
+    return ret;
+}
+
+vehicle_profile vehicle::autodrive_controller::compute_profile( orientation facing ) const
+{
+    const tileray tdir( units::from_degrees( facing * TURNING_INCREMENT ) );
+    std::map<int, std::pair<int, int>> extent_map;
+    int y_min = std::numeric_limits<int>::max();
+    const point pivot = driven_veh.pivot_point();
+    for( const vehicle_part &part : driven_veh.parts ) {
+        if( part.removed ) {
+            continue;
+        }
+        tripoint pos;
+        vehicle::coord_translate( tdir, pivot, part.mount, pos );
+        if( extent_map.find( pos.y ) == extent_map.end() ) {
+            extent_map[pos.y] = { pos.x, pos.x };
+            y_min = std::min( y_min, pos.y );
+        } else {
+            auto &extent = extent_map[pos.y];
+            extent.first = std::min( extent.first, pos.x );
+            extent.second = std::max( extent.second, pos.x );
+        }
+    }
+    // TODO: find area reachable by rotor blades
+    vehicle_profile ret;
+    ret.tdir = tdir;
+    ret.y_min = y_min;
+    ret.y_max = y_min + extent_map.size() - 1;
+    ret.x_extents.resize( extent_map.size() );
+    for( int y = y_min; y <= ret.y_max; y++ ) {
+        ret.x_extents[y - y_min] = extent_map[y];
+    }
+    return ret;
+}
+
+
+bool vehicle::autodrive_controller::check_obstacle( tripoint pt ) const
+{
+    const map &here = get_map();
+    const optional_vpart_position ovp = here.veh_at( pt );
+    if( ovp ) {
+        return &ovp->vehicle() != &driven_veh;
+    }
+    if( g->critter_at( pt, true ) ) {
+        return true;
+    }
+    const furn_id furniture = here.furn( pt );
+    if( furniture != f_null && furniture.obj().movecost != 0 ) {
+        return true;
+    }
+    const ter_id terrain = here.ter( pt );
+    if( terrain == t_null ) {
+        return true;
+    }
+    if( terrain == t_open_air ) {
+        return !data.air_ok;
+    }
+    const ter_t &terrain_type = terrain.obj();
+    if( data.water_ok && terrain_type.has_flag( TFLAG_SWIMMABLE )) {
+        return false;
+    }
+    if( terrain_type.movecost <= 0 ) {
+        return true;
+    } else if( terrain_type.movecost == 2 || terrain_type.has_flag( "NOCOLLIDE" ) ) {
+        return !data.land_ok;
+    } else if( terrain_type.bash.str_max >= 0 && !terrain_type.bash.bash_below ) {
+        return true;
```

Comment:
The logic here could use some comments.

Reply:
Added comments.

Reply:
I see you also simplified the (!)data.land_ok logic; thanks!

Reply:
I've also renamed it to check_drivable and inverted the return values. Hopefully that improves readability further.